### DISCUSSION TITLE
Plot circle in 3d space

### DISCUSSION THREAD

**Discussion by platypus999** [cetz-package/cetz#749] [Q&A] [answered]
I want to draw a circle representing a coordinate point in a 3D space within an `ortho` environment. However, with the code below, the circle appears to be slanted, as seen in the image. Is there a way to rotate the circle so that its normal vector is parallel to the camera's line of sight?

```typ
#cetz.canvas({
  import cetz.draw: *
  ortho(x: 20deg, y: 30deg, {
    on-xz({
      grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
      line((0, 0), (0, 5.5), mark: (end: ">"))
      line((0, 0), (5.5, 0), mark: (end: ">"))
    })
    on-xy({
      grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
      line((0, 0), (0, 5.5), mark: (end: ">"))
    })
    on-yz({
      grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
    })
    circle((3,3,4), radius: 0.2, fill:red)
  })
})
```

![image](https://github.com/user-attachments/assets/62aa9bac-308b-4738-84f8-6a7daa36edbc)

**Comment by platypus999** [comment]
In the meantime, I have found a workaround to this problem, by giving up on `ortho` and using the `persp` function introduced in https://github.com/cetz-package/cetz/discussions/746#discussioncomment-11220430.
This works because the direction of the x and y axes does not change after transformation.

```typ
#cetz.canvas({
  import cetz.draw: *
  let persp(a, c: calc.sqrt(2)/2) = {
    set-transform(
      ((1, 0, calc.cos(a) * c,0),
       (0,-1,-calc.sin(a) * c,0),
       (0, 0, 0,  0),
       (0, 0, 0,  1)))
  }
  persp(45deg, c: -0.5)
  on-xz({
    grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
    line((0, 0), (0, 5.5), mark: (end: ">"))
    content((0, 5.8), $z$)
    line((0, 0), (5.5, 0), mark: (end: ">"))
    content((5.7, 0), $x$)
  })
  on-xy({
    grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
    line((0, 0), (0, 5.5), mark: (end: ">"))
    content((0, 5.8), $y$)
  })
  on-yz({
    grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
  })
  circle((3,3,4), radius: 0.2, fill:red)
})
```

 
![image](https://github.com/user-attachments/assets/e3e34544-8bbe-45af-aef5-3250a384452a)

However, I still do not know how to plot unslanted circles in the `ortho` environment, or any other 3d space of arbitrary axis directions.

**Comment by platypus999** [comment, marked as answer]
Upon further research (and trial-and-error), the easiest way I could find to plot an unslanted circle in the `ortho` environment was to put the circle in another `cetz.canvas`, and encapsulate it with `content`.

So instead of `circle(coordinate, ...)`, I will do `content(coordinate, [#cetz.canvas({circle((), ...)})])`.
This will work with any shape (or combination of shapes that can be contained in a canvas).

I will mark this question as answered, but any further insights / better implementation suggestions are welcome!

```typ
#cetz.canvas({
  import cetz.draw: *
  ortho(x: 20deg, y: 30deg, {
    on-xz({
      grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
      line((0, 0), (0, 5.5), mark: (end: ">"), name:"line")
      line((0, 0), (5.5, 0), mark: (end: ">"), name:"line")
    })
    on-xy({
      grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
      line((0, 0), (0, 5.5), mark: (end: ">"))
    })
    on-yz({
      grid((0, 0), (5.5, 5.5), step: 1.0, stroke: gray + 0.2pt)
    })
    content((3,3,4), [#cetz.canvas({circle((), radius: 0.2, fill:red)})])
  })
})
```

![image](https://github.com/user-attachments/assets/d922240f-6d96-4569-b61b-ac335a022075)

**Reply by johannes-wolf** [reply]
You could also use an `anchor` inside the `ortho` env, and draw the circle outsides.
